**1. Import Pandas and Read the csv file.**

# Ecommerce Purchases Exercise Overview

Hi Guys,<br>
After a crash course on pandas for data analysis, its time to do some practice!!!<br>
Because of privacy issues, I have created a fake dataset here with 30K entries. <br>
The situation is, customers are providing some personal information while purchasing stuff on-line or in-store. For some reasons, your client wants to know the answer to some of his questions from the dataset, let's try to help him!.<br>

*Feel free to consult the solutions if needed. Please note, the tasks given in the exercises, can be solved in different ways. Try your best answer and compare with the solutions.* <br>
&#9758; The dataset is provided in the course downloads. 

In [1]:
import pandas as pd

df = pd.read_csv('Cust_Purch_FakeData.csv')

print(df)

      prefix    first       last              email  gender  age  \
0        Dr.      Ray     Morton    sebvajom@kol.km    Male   38   
1       Miss  Claudia  Rodriquez      acu@jatsot.ug  Female   51   
2       Miss    Harry      Meyer          zuz@lo.wf  Female   51   
3       Miss    Edith    Gilbert  hansohsi@jupec.md  Female   55   
4        Dr.     Lura     Murphy     webediti@je.st  Female   20   
...      ...      ...        ...                ...     ...  ...   
29995    Mr.     John    Maxwell  larlisu@sejenu.ad  Female   38   
29996    Dr.   Alfred       Bass      le@duridzo.gh  Female   24   
29997   Mrs.   Edward     Curtis      ta@kucemah.et  Female   56   
29998    Mr.      Max      Woods       li@mincoz.mp  Female   64   
29999    Mr.     Olga  Rodriguez         ner@ota.ec    Male   22   

                             company                profession  \
0                     Medtronic Inc.          Health Therapist   
1       Ames Department Stores, Inc.          Healt

**2. Its good idea to see how the data look like, display first 5 rows of your data-set.**

In [2]:
df.head()

prefix    first       last              email  gender  age  \
0    Dr.      Ray     Morton    sebvajom@kol.km    Male   38   
1   Miss  Claudia  Rodriquez      acu@jatsot.ug  Female   51   
2   Miss    Harry      Meyer          zuz@lo.wf  Female   51   
3   Miss    Edith    Gilbert  hansohsi@jupec.md  Female   55   
4    Dr.     Lura     Murphy     webediti@je.st  Female   20   

                        company              profession           phone  \
0                Medtronic Inc.        Health Therapist  (987) 619-2695   
1  Ames Department Stores, Inc.        Health Therapist  (356) 736-7352   
2                     CSX Corp.     Political Scientist  (539) 246-1806   
3        Murphy Oil Corporation  Transportation Manager  (984) 962-7494   
4                  PETsMART Inc            Statistician  (902) 568-9748   

    postal province             cc_no   cc_exp                    cc_type  \
0  B6V 3W3       MB  5020000000000230  05/2018                       Solo   
1  G7M 5F3       SK  5020000000000230  07/2028                       Visa   
2  A0Z 6P9       NS  6300000000000000  02/2023                     Switch   
3  P9I 9H3       YT  3530000000000000  02/2028                    Maestro   
4  S1A 6K0       ON  4030000000000000  10/2025  Diners Club International   

   price(CAD) fav_color              ip    weekday ampm        date  
0        8.36      Blue    126.23.139.2     Sunday   pm  04/05/1930  
1       68.31     Black  106.198.76.211    Tuesday   am  12/20/1926  
2       34.65     Black  186.150.187.29  Wednesday   pm  08/20/1931  
3       64.59     White   80.140.57.161   Saturday   am  06/18/2001  
4       20.83    Yellow   211.103.43.41     Friday   pm  06/14/2045

**3. How many entries your data have? 
Can you tell the no. of columns in your data?**

In [3]:
print(df.shape)

print(len(df.columns))

(30000, 20)
20


**4. What are the max and min ages of your customer? Can you find mean of your customer?**

In [4]:
print(df.columns)

max_age = df['age'].max()
min_age = df['age'].min()
mean_age = df['age'].mean()

print("Maximum age:", max_age)
print("Minimum age:", min_age)
print("Average age:", mean_age)


Index(['prefix', 'first', 'last', 'email', 'gender', 'age', 'company',
       'profession', 'phone', 'postal', 'province', 'cc_no', 'cc_exp',
       'cc_type', 'price(CAD)', 'fav_color', 'ip', 'weekday', 'ampm', 'date'],
      dtype='object')
Maximum age: 65
Minimum age: 18
Average age: 41.550066666666666


**5. What are the three most common customer's names?**<br>
&#9989; [<code>**value_counts()**</code>](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html) returns object containing counts of unique values. The resulting object will be in descending order so that the first element is the most frequently-occurring element. Excludes NA values by default.

In [5]:
print(df.columns)
top_names = df['first'].value_counts().head(3)
print(top_names)



Index(['prefix', 'first', 'last', 'email', 'gender', 'age', 'company',
       'profession', 'phone', 'postal', 'province', 'cc_no', 'cc_exp',
       'cc_type', 'price(CAD)', 'fav_color', 'ip', 'weekday', 'ampm', 'date'],
      dtype='object')
first
Willie     130
Francis    124
Eula        86
Name: count, dtype: int64


**6. Two customers have the same phone number, can you find those customers**?

In [6]:
duplicate_phones = df['phone'].value_counts()

duplicate_phones = duplicate_phones[duplicate_phones > 1].index

customers_with_duplicate_phones = df[df['phone'].isin(duplicate_phones)]

print(customers_with_duplicate_phones)


   prefix  first   last                email  gender  age            company  \
15   Mrs.  Lilly  Tyler      kofadu@itohi.tf  Female   38          CSX Corp.   
16   Mrs.  Peter   Cain  megkosig@anazeor.gn    Male   27  Campbell Soup Co.   

             profession           phone   postal province             cc_no  \
15  Structural Engineer  (263) 382-8004  V7K 1E3       ON    30000000000000   
16   Insurance Adjuster  (263) 382-8004  E8T 2B4       YT  3530000000000000   

     cc_exp                    cc_type  price(CAD) fav_color              ip  \
15  11/2022  Diners Club Carte Blanche        9.61    Yellow   74.124.37.227   
16  03/2024                       Solo       13.74     Black  25.207.141.135   

     weekday ampm        date  
15  Saturday   am  03/30/1985  
16   Tuesday   am  08/05/1950  


In [7]:
# lets find what phone number is twice
phone_counts = df['phone'].value_counts()

phones_used_twice = phone_counts[phone_counts == 2].index

print("Phone numbers used exactly twice:")
print(phones_used_twice)


Phone numbers used exactly twice:
Index(['(263) 382-8004'], dtype='object', name='phone')


In [8]:
# Code here please

In [9]:
# Now we know the phone number, let's find out the other stuff!


**7. How many customers have profession "Structural Engineer"?**

In [10]:
count_structural_engineers = df[df['profession'] == 'Structural Engineer'].shape[0]

print("Number of Structural Engineers:", count_structural_engineers)


Number of Structural Engineers: 87


**8. How many male customers are 'Structural Engineer'?**

In [11]:
male_structural_engineers = df[
    (df['gender'] == 'Male') & (df['profession'] == 'Structural Engineer')
]
count = male_structural_engineers.shape[0]

print("Number of male Structural Engineers:", count)


Number of male Structural Engineers: 43


**9. Find out the female Structural Engineers from province Alberta (AB)?** 

In [12]:
male_structural_engineers = df[
    (df['gender'] == 'Female') & (df['profession'] == 'Structural Engineer')
]
count = male_structural_engineers.shape[0]

print("Number of female Structural Engineers:", count)


Number of female Structural Engineers: 44


**10. What is the max, min and average spending?**

In [13]:
# Assuming the spending data is in a column named 'Spending'
max_spending = df['price(CAD)'].max()
min_spending = df['price(CAD)'].min()
avg_spending = df['price(CAD)'].mean()

print("Max. spending:", max_spending)
print("Min. spending:", min_spending)
print("Avg. spending:", avg_spending)


Max. spending: 100.0
Min. spending: 0.0
Avg. spending: 49.990775


**11. Who did not spend anything? Company wants to send a deal to encourage the customer to buy stuff!**

In [14]:
no_spending_customers = df[df['price(CAD)'] == 0]

print(no_spending_customers)


      prefix  first   last           email  gender  age              company  \
5320     Dr.  Bruce  Bryan  ru@pubuspuh.cl    Male   59  Wal-Mart Stores Inc   
10597   Mrs.  Flora  Clark      wad@me.com  Female   27         MetLife Inc.   

                profession           phone   postal province            cc_no  \
5320   Engineer Technician  (709) 446-8317  H6H 3Y0       NU  201000000000000   
10597      Cruise Director  (775) 373-6590  B2U 8K6       NB   30100000000000   

        cc_exp                    cc_type  price(CAD) fav_color  \
5320   04/2024                   Bankcard         0.0     White   
10597  05/2020  Diners Club Carte Blanche         0.0       Red   

                  ip    weekday ampm        date  
5320     82.70.62.64  Wednesday   am  12/11/1900  
10597  231.156.15.63   Thursday   pm  08/22/1904  


**12. As a loyalty reward, company wants to send thanks coupon to those who spent 100CAD or more, please find out the customers?**

In [15]:
loyal_customers = df[df['price(CAD)'] >= 100]

print(loyal_customers)


      prefix    first         last          email  gender  age  \
76      Mrs.  Gregory        Brown     hav@jek.gs  Female   31   
21093   Mrs.     Cody  Christensen    get@jovu.ag    Male   28   
24385   Miss   Lizzie        Dixon  goh@tuwjaz.gd  Female   38   

                         company              profession           phone  \
76           E*Trade Group, Inc.                Novelist  (625) 537-8923   
21093        National City Corp.  Hospital Administrator  (261) 737-3292   
24385  FleetBoston Financial Co.    Compensation Analyst  (989) 239-1752   

        postal province             cc_no   cc_exp   cc_type  price(CAD)  \
76     X3S 4Q2       PE  6010000000000000  04/2018      Visa       100.0   
21093  X0S 1O5       PE   201000000000000  03/2024       JCB       100.0   
24385  V8X 9V6       NB  6300000000000000  07/2019  Bankcard       100.0   

      fav_color             ip    weekday ampm        date  
76        Green  212.100.18.95     Monday   am  10/23/2063  
210

**13. How many emails are associated with this credit card number '5020000000000230'?**

In [16]:
target_card = '502000000000230'
associated_emails = df[df['cc_no'] == target_card]['email']

# Count the number of unique emails
email_count = associated_emails.nunique()

print(f"Total unique emails linked to {target_card}: {email_count}")


Total unique emails linked to 502000000000230: 0


**14. We need to send new cards to the customers well before the expire, how many cards are expiring in 2019?**<br>
*Use `sum()` and `count()` and see the difference in their use :)*

In [17]:
df['cc_exp'] = pd.to_datetime(df['cc_exp'])

expiring_2019 = df[df['cc_exp'].dt.year == 2019]

print("Using count():", expiring_2019['cc_exp'].count())

print("Using sum():", (df['cc_exp'].dt.year == 2019).sum())


Using count(): 2684
Using sum(): 2684


C:\Users\FLS\AppData\Local\Temp\ipykernel_684\113932707.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['cc_exp'] = pd.to_datetime(df['cc_exp'])


In [18]:
#cust[cust['cc_exp'].apply(lambda x: x[5:]) == '19'].count()


**15. How many people use Visa as their Credit Card Provider?**

**16. Can you find the customer who spent 100 CAD using Visa?**

In [19]:
# Code here please

**17. What are two most common professions?**

In [20]:
# Code here please

**18. Can you tell the top 5 most popular email providers? (e.g. gmail.com, yahoo.com, etc...)**

In [21]:
# Code here please

**19. Is there any customer who is using email with "am.edu"?**<br>
Hint: Use `lambda` expression in `apply()`. split the email address at `@`.  

In [22]:
# Code here please

**20. Which day of the week, the store gets more customers?**

In [23]:
# Code here please

# Excellent work!